2018 Robot/src/org/usfirst/frc/team4999/robot/sensors/VMXPi.java

```diff
@@ -0,0 +1,121 @@
+package org.usfirst.frc.team4999.robot.sensors;
+
+import java.io.IOException;
+import java.net.DatagramPacket;
+import java.net.DatagramSocket;
+import java.nio.ByteBuffer;
+
+import edu.wpi.first.wpilibj.PIDSource;
+import edu.wpi.first.wpilibj.PIDSourceType;
+
+/**
+ *
+ */
+public class VMXPi implements PIDSource {
+
+    private static int PORT = 5800;
+	private DatagramSocket server;
+	DatagramPacket packet = new DatagramPacket(new byte[17], 17);
+	
+	private Thread recieveLoop;
+	public static final long RESPONSECUTOFF = 200;
+	private double zAngle, zRate;
+	
+	// When two synchronized blocks are synchronized on the same object, only one block is allowed to run at a time.
+	// By synchronizing before we read and write the zAngle and zRates, we can make sure they don't change while we're reading them
+	private Object syncLock = new Object();
+	
+	private PIDSourceType m_source = PIDSourceType.kDisplacement;
+	
+	private long millis;
+    
+	private static VMXPi instance;
+	
+	public static VMXPi getInstance() {
+		if(instance == null) {
+			instance = new VMXPi();
+		}
+		return instance;
+	}
+	
+	private VMXPi() {
+		try {
+			server = new DatagramSocket(PORT);
+		} catch (IOException e) {
+			e.printStackTrace();
+		}
+		recieveLoop = new Thread() {
+			ByteBuffer buff = ByteBuffer.allocate(17);
+			
+			@Override
+			public void run() {
+				setName("VMX Thread");
+				while(!Thread.interrupted()) {
+					try {
+						server.receive(packet);
+						byte[] data = packet.getData();
+						int checksum = 0;
+						for(int i = 1; i < data.length; i++) {
+							checksum += (data[i]&0xFF);
+						}
+						if(checksum % 255 == (data[0]&0xFF)) { // 0xFF removes the sign
+							buff.rewind();
+							buff.put(data);
+							synchronized(syncLock) {
+								zAngle = buff.getDouble(1);
+								zRate = buff.getDouble(9);
+								millis = System.currentTimeMillis();
+							}
+						} else System.out.format("%d != %d\n", checksum % 255, (data[0]&0xFF));
```

Comment:
Same comment as above.

Reply:
Fixed in https://github.com/momentumfrc/2018PowerUp/commit/e0f14dc083eb338f48f8450292b003332916b24f